File: leetcode/problems/language-specific/go/20.valid-parentheses.go
/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */

// @lc code=start
func isValid(s string) bool {
	if len(s) <= 1 {
		return false
	}

	stack := []byte{}
	for i := range s {
		if s[i] == '(' || s[i] == '{' || s[i] == '[' {
			stack = append(stack, s[i])
		} else {
			if len(stack) <= 0 ||
				(s[i] == ')' && stack[len(stack)-1] != '(') ||
				(s[i] == '}' && stack[len(stack)-1] != '{') ||
				(s[i] == ']' && stack[len(stack)-1] != '[') {
				return false
			}

			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}

// @lc code=end

File: leetcode/problems/language-specific/python/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start

# * Iterative Stack Solution | O(n) Time | O(n) Space
# * n -> The length of s string


class Solution:
    def isValid(self, s: str) -> bool:
        s_stack = []
        for char in s:
            if char == "(" or char == "{" or char == "[":
                s_stack.append(char)
            else:
                if (
                    not s_stack
                    or (char == ")" and not s_stack[-1] == "(")
                    or (char == "}" and not s_stack[-1] == "{")
                    or (char == "]" and not s_stack[-1] == "[")
                ):
                    return False

                s_stack.pop()

        return not s_stack


# @lc code=end
